crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2028,118 @@ impl<'db> Class<'db> {
         self.iter_mro(db).contains(&ClassBase::Class(other))
     }
 
+    /// Return the explicit `metaclass` of this class, if one is defined.
+    ///
+    /// ## Note
+    /// Only call this function from queries in the same file or your
+    /// query depends on the AST of another file (bad!).
+    fn explicit_metaclass(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let class_stmt = self.node(db);
+        let metaclass_node = &class_stmt
+            .arguments
+            .as_ref()?
+            .find_keyword("metaclass")?
+            .value;
+        Some(if class_stmt.type_params.is_some() {
+            // when we have a specialized scope, we'll look up the inference
+            // within that scope
+            let model = SemanticModel::new(db, self.file(db));
+            metaclass_node.ty(&model)
+        } else {
+            // Otherwise, we can do the lookup based on the definition scope
+            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
+            definition_expression_ty(db, class_definition, metaclass_node)
+        })
+    }
+
+    /// Return the metaclass of this class, or `Unknown` if the metaclass cannot be inferred.
+    pub(crate) fn metaclass(self, db: &'db dyn Db) -> Type<'db> {
+        // TODO: `type[Unknown]` would be a more precise fallback
+        // (needs support for <https://docs.python.org/3/library/typing.html#the-type-of-class-objects>)
+        self.try_metaclass(db).unwrap_or(Type::Unknown)
+    }
+
+    /// Return the metaclass of this class, or an error if the metaclass cannot be inferred.
+    #[salsa::tracked]
+    pub(crate) fn try_metaclass(self, db: &'db dyn Db) -> Result<Type<'db>, MetaclassError<'db>> {
+        /// Infer the metaclass of a class, tracking the classes that have been visited to detect
+        /// cyclic definitions.
+        fn infer<'db>(
+            db: &'db dyn Db,
+            class: Class<'db>,
+            seen: &mut SeenSet<Class<'db>>,
+        ) -> Result<Type<'db>, MetaclassError<'db>> {
+            let mut base_classes = class
+                .explicit_bases(db)
+                .iter()
+                .copied()
+                .filter_map(Type::into_class_literal);
```

Comment:
I think we should add a test where one of the bases is `Unknown`, but I think the current behavior here is already mostly correct; we should (mostly) effectively ignore that base when it comes to metaclass calculation. If there are other bases, we should assume the metaclass of the unknown base is compatible with anything, and if the only base is `Unknown`, we should assume its metaclass is `type`.

The reason I say "mostly"  above is that technically if we have at least one Unknown (or Any or Todo) base, then the returned metaclass should be a `type[]` type (instead of a class literal type), and should be intersected with `Any` or `Unknown` or `Todo`. So:

```
from nonexistent_module import UnknownClass  # error: [unresolved-import]

class C(UnknownClass): ...
reveal_type(C.__class__)  # revealed: type[type] & Unknown

class M(type): ...
class A(metaclass=M): ...
class B(A, UnknownClass): ...

reveal_type(B.__class__)  # revealed: type[M] | Unknown
```

This is a bit subtle, but the reason for the intersection with Unknown in these cases is that the unknown base _might_ have a metaclass that is a subclass of `type` (in the first case) or `M` (in the second case). We reflect this possibility by intersecting with Unknown. `T & Unknown` effectively means "this type may be something narrower than T, but it can't be anything not in T".

It may seem like `type[M] & Unknown` is no different from `type[M]`, because `type[M]` is already a type that includes all subclasses of M! But the difference is that `Unknown/Any` are forgiving; they always assume the most compatible type. If you ask whether `type[M]` is a subclass of `type[N]`, the answer may be "no" -- if you ask whether `type[M] & Unknown` is a subclass of `type[N]`, the answer is always yes.

But all that said, this requires a `type[]` type, which we don't have yet, and it's not high priority, so I think we should just TODO it for this PR. I'd still want to add the above tests, but the test would reveal e.g. `Literal[M]`, with a TODO comment right above the `revealed: ` line that it should be `type[M] & Unknown` instead.)